src/compiler/moduleNameResolver.ts

```diff
@@ -738,6 +741,24 @@ namespace ts {
         }
     }
 
+    function tryLoadModuleUsingPathsIfEligible(extensions: Extensions, moduleName: string, loader: ResolutionKindSpecificLoader, state: ModuleResolutionState) {
+        const { baseUrl, paths } = state.compilerOptions;
+        if (baseUrl && paths && isEligibleForPathMapping(moduleName, paths)) {
+            if (state.traceEnabled) trace(state.host, Diagnostics.paths_option_is_specified_looking_for_a_pattern_to_match_module_name_0, moduleName);
+            return tryLoadModuleUsingPaths(extensions, moduleName, baseUrl, paths, loader, /*onlyRecordFailures*/ false, state);
+        }
+    }
+
+    function isEligibleForPathMapping(moduleName: string, paths: MapLike<string[]>): boolean {
+        if (isRootedDiskPath(moduleName) || isUrl(moduleName)) {
+            // check for potential alias otherwise { '*': './src/*' } will match
+            const root = moduleName.slice(0, getRootLength(moduleName));
+            const potentialAlias = getOwnKeys(paths).find(path => path.startsWith(root));

```

Comment:
`find` and `startsWith` are ES2015 and we support ES5 environments, so use the helper functions from `core.ts`